### DISCUSSION TITLE
Graphql for dummies...

### DISCUSSION THREAD

**Discussion by ankyfive** [requarks/wiki#2757] [Help / Questions] [answered]
**Question**
Hello,
I am new to graphql and it is quite difficult for me.
In order to understand better how it works, could someone give me an example for searching pages and for moving it just to see how to whrite the querries ?
Thanks in advance

**Host Info (please complete the following information):**
OS: Debian 9.12
Wiki.js version: 2.5.159
Database engine:postgres 9.6.17

**Comment by Smankusors** [comment]
[this might be a good starting point](https://graphql.org/learn/)

**Comment by ankyfive** [comment]
Thanks :-)

**Comment by ankyfive** [comment]
I tried this but don't understand why it does not work : 

query{
  pages : PageQuery {
    list ( limit : 10)
  }
}

It is just a question of what is the syntax of this langage. For me the doc is not clear, or has not the same syntax as your app

**Comment by ankyfive** [comment]
I found that : 
query{pages{list{title}}} 

But I don't undestand your doc...

**Comment by ankyfive** [comment]
If someone can help, I can't spend hours to learn this langage I am not developper.

I just would like to know how to delete a page I tried : 
mutation{pages{delete(id:11)}}
But does nothing

Thanks for any help

**Comment by Smankusors** [comment]
indeed there's ~no existing~ little docs for Wiki.js API

perhaps you can use your DevTools and see how the Wiki.js frontend send some queries to the GraphQL API

**Comment by winkelement** [comment]
For docs check https://docs.requarks.io/dev/api

To delete a page (and get back all possible response results):

```
mutation {
  pages {
    delete(id: 11) {
      responseResult {
        succeeded
        errorCode
        slug
        message
      }
    }
  }
}

```

**Reply by JCOvergaar** [reply]
that's a list of error messages, not documentation.

**Comment by ankyfive** [comment]
Thanks but it not working, I am connected as administrator and I have the following message : 
 errors": [
    {
      "message": "Forbidden",
      "locations": [

**Comment by winkelement** [comment, marked as answer]
Make sure you have created a valid API Token for Authorization as described here

> For docs check https://docs.requarks.io/dev/api

Send the API Token in HTTP Header like this with your requests:

```
{
  "Authorization": "Bearer ...LEfVqGMYQiabK167dt_0Q"
}
```

Check out the GraphQL Playground tool via Browser at `/graphql`

<img width="1297" alt="Bildschirmfoto 2020-11-26 um 14 46 22" src="https://user-images.githubusercontent.com/3525191/100358356-605b7980-2ff6-11eb-8472-cccf1bdf0682.png">

**Comment by dazzag24** [comment]
Example code in Python to list, search and create a page in wiki.js

```
import requests
import json
import os
from dotenv import load_dotenv

load_dotenv()

WIKIJS_URI = os.getenv("WIKIJS_URI")
WIKIJS_TOKEN =  os.getenv("WIKIJS_API_TOKEN")

wikijsHeaders = {"Authorization": "Bearer " + WIKIJS_TOKEN}
wikijsStatusCode = 200

wikijsPageListQuery = """
query{
  pages {
    list(limit:10 ){
      title,
      path,
      id
    }
  }
}
"""

wikijsPageSearchQuery = """
query {
  pages {
    search(query: "Badger") {
      results {
          id,
          title,
          description,
          path,
          locale
      },
      suggestions,
      totalHits
    }
  }
}
"""

wikijsCreatePageMutation = """
mutation {
  pages {
    create (content: "# Test content. _Foo_ Bar Badger",
    description:"This is a description",
    editor: "markdown",
    isPublished: true,
    isPrivate: false,
    locale: "en",
    path: "/Development/Test/TestMcTestfaceIII",
    tags: ["tag1", "tag2"],
    title: "This is a title") {
      responseResult {
        succeeded,
        errorCode,
        slug,
        message
      },
      page {
        id,
        path,
        title
      }
    }
  }
}
"""

def wikijs_graphql_call(uri, query, statusCode, headers):
    request = requests.post(uri, json={'query': query}, headers=headers)
    if request.status_code == statusCode:
        return request.json()
    else:
        print(request.content)
        raise Exception(f"Unexpected status code returned: {request.status_code}")

print("Send a list query")
result = wikijs_graphql_call(WIKIJS_URI, wikijsPageListQuery, wikijsStatusCode, wikijsHeaders)
print(json.dumps(result, indent=4))

print("\nSend a page mutation to add a page")
result = wikijs_graphql_call(WIKIJS_URI, wikijsCreatePageMutation, wikijsStatusCode, wikijsHeaders)
print(json.dumps(result, indent=4))

print("\nSend a search query")
results = wikijs_graphql_call(WIKIJS_URI, wikijsPageSearchQuery, wikijsStatusCode, wikijsHeaders)
#print(json.dumps(results, indent=4))
print("Found {} result(s)\n".format(results['data']['pages']['search']['totalHits']))
for hit in results['data']['pages']['search']['results']:
  print(json.dumps(hit, indent=4))
```

URL and API token are read from a .env file.  Needs content like so:

```
WIKIJS_URI = 'https://mywiki.serverhere.com/graphql'
WIKIJS_API_TOKEN = '<API_TOKEN_OBTAINED_FROM_THE_WIKIJS_ADMIN_PAGES_HERE>'
```

**Reply by JCOvergaar** [reply]
thank you!

**Comment by ankyfive** [comment]
Thanks for the info :-)
